pkg/monitor/health/cheker.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2025 the Unikorn Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package health
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+
+	unikornv1 "github.com/unikorn-cloud/core/pkg/apis/unikorn/v1alpha1"
+	"github.com/unikorn-cloud/core/pkg/cd"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrTypeConversion = errors.New("type conversion error")
+)
+
+// Lister is a generic inteface for operating on and iterating over resource
+// lists (as no such interface is required by apimachinery.
+type Lister[T unikornv1.ManagableResourceInterface] interface {
+	client.ObjectList
+	All() iter.Seq[T]
+}
+
+// Checker lists all resources of the specified type and does a health check on it.
+// The type itself is constrained to a manageable resource so we can get the label selector
+// to pass to the CD layer to get all applications for the resource, then once we have
+// checked the status of those applications we can set the condition generically, again
+// as provided by the manageable resource interface.
+type Checker[T unikornv1.ManagableResourceInterface, L Lister[T]] struct {
+	// client allows access to Kubernetes resources.
+	client client.Client
+	// driver is the CD driver.
+	driver cd.Driver
+	// t is storage for the manageable resource.
```

Comment:
This isn't used; `T` is now just in the signature so the Lister type can have a type parameter.

Reply:
Ah true, we just directly reference the list storage, pew pew

Reply:
I am pleasantly surprised that type inference just work!